ENH Poisson loss for HistGradientBoostingRegressor

lorentzenchr opened the pull request:
#### Reference Issues/PRs
This PR partly addresses #16668 and #5975.

#### What does this implement/fix? Explain your changes.
This PR implements the Poisson loss for `HistGradientBoostingRegressor`, i.e. splitting based on improvement in Poisson deviance.

Commit:
PEP8 blank line after class docstring

Commit:
ENH add Poisson loss class to histogram gradient boosting

Commit:
TST add tests for Poisson loss class

Commit:
BUG hessians are not constant for Poisson loss

Commit:
ENH add Poisson loss to HistGradientBoostingRegressor

Commit:
TST test Poisson loss for HistGradientBoostingRegressor

lorentzenchr commented:
ping @NicolasHug

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -211,11 +218,33 @@ def test_baseline_categorical_crossentropy():
+    loss = _LOSSES['poisson_loss'](sample_weight=None)
+    y_train = rng.poisson(size=100).astype(np.float64)
+    # Make sure at least one sample point is larger than zero
+    y_train[0] = 1.
```
```suggestion
    # Sanity check, make sure at least one sample is non-zero so we
    # don't take log(0)
    assert y_train.sum() > 0
```

NicolasHug replied:
Though TBF I'm fine removing this given the very low probability that this ever happens (basically if it passes all the CIs, it's fine)

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -211,11 +218,33 @@ def test_baseline_categorical_crossentropy():
+    assert_all_finite(baseline_prediction)
+    # Make sure baseline prediction produces the mean of all targets
+    y_baseline = loss.inverse_link_function(baseline_prediction)
+    assert_almost_equal(np.mean(y_baseline), y_train.mean())
```
For consistency with the other checks we should directly compare `baseline_prediction`

```suggestion
    # Make sure baseline prediction produces the log of the mean of all targets
    assert_almost_equal(np.log(y_train.mean()), baseline_prediction)
```

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -211,11 +218,33 @@ def test_baseline_categorical_crossentropy():
+    assert_almost_equal(np.mean(y_baseline), y_train.mean())
+
+    # Test baseline for y_true = 0
+    y_train.fill(0.)
```
```suggestion
    y_train = np.zeros(100)
```

lorentzenchr replied:
Why do you prefer it this way?

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -192,6 +192,24 @@ def test_least_absolute_deviation():
     assert gbdt.score(X, y) > .9
 
 
+def test_poisson_loss():
```
Would it make sense to also test that the score is above a given threshold?

lorentzenchr replied:
Unlike r2 score, it is hard to give an absolute "good" value for the poisson deviance. With the "d2 score" of #15244 this would make more sense.
I added a `DummyRegressor` with mean as prediction. This is (almost) equivalent to a d2 score. And I added out-of-sample tests.

NicolasHug reviewed (commented):
Thanks @lorentzenchr , this looks good!

I mostly have minor comments.

Should we check that we always have y >= 0?

Please make a minor update to `ensemble.rst` (around line 955) to document the new loss.
This will also need an entry in the what's new

Commit:
MNT rename poisson_loss to poisson

Commit:
TST improve test_poisson for hist gradient boosting

lorentzenchr commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -76,17 +79,19 @@ def fprime(x):
 
-    optimum = newton(func, x0=x0, fprime=fprime, fprime2=fprime2)
+    optimum = newton(func, x0=x0, fprime=fprime, fprime2=fprime2,
+                     maxiter=70, tol=2e-8)
```
RuntimeError: Failed to converge after 50 iterations

thomasjpfan replied:
It seems like, we can increase `max_iter` to 120 and not have to add `tol`.

Commit:
MNT move up poisson loss to regression losses

Commit:
TST sanity check that sum(y)>0 for poisson

Commit:
TST compare poisson baseline_prediction directly

Commit:
ENH check for invalid y for poisson in hist gradient boosting

Commit:
DOC add poisson loss to user guide

Commit:
DOC mention log-link and y>=0 for poisson loss in hist gradient boosting

Commit:
DOC whats new

lorentzenchr commented:
@NicolasHug Thanks for your fast first review pass. I think I addressed all comments. I have to say, the histogram gradient boosting implementation seems like a piece of art. I wish it would have been that easy to include Poisson for linear models 😄

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/loss.py:
```diff
@@ -269,6 +273,62 @@ def update_leaves_values(self, grower, y_true, raw_predictions,
+    def __init__(self, sample_weight):
+        # If sample weights are provided, the hessians and gradients
+        # are multiplied by sample_weight, which means the hessians are
+        # equal to sample weights.
```
The comment was meant for e.g LS because normally LS has constant hessians (equal to 1), but they're not constant anymore when we have SW (where hessian = SW). But this doesn't apply for poisson since the hessians are never constant anyway, so let's remvoe the comment

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -192,6 +195,56 @@ def test_least_absolute_deviation():
+    'y',
+    [([1., -2., 0.]),
+     ([0., 0., 0.]),
+     ])
```
That might fit in one line?

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -192,6 +195,56 @@ def test_least_absolute_deviation():
+                             random_state=rng)
+    # We create a log-linear Poisson model and downscale coef as it will get
+    # exponentiated.
+    coef = rng.uniform(low=-2., high=+2., size=n_features) / np.amax(X, axis=0)
```
```suggestion
    coef = rng.uniform(low=-2, high=2, size=n_features) / np.max(X, axis=0)
```

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -192,6 +195,56 @@ def test_least_absolute_deviation():
+    coef = rng.uniform(low=-2., high=+2., size=n_features) / np.amax(X, axis=0)
+    y = rng.poisson(lam=np.exp(X @ coef))
+    X_train, X_test = X[0:n_train, :], X[n_train:n_train+n_test, :]
+    y_train, y_test = y[0:n_train], y[n_train:n_train+n_test]
```
you can use `train_test_split()`

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -192,6 +195,56 @@ def test_least_absolute_deviation():
+    y = rng.poisson(lam=np.exp(X @ coef))
+    X_train, X_test = X[0:n_train, :], X[n_train:n_train+n_test, :]
+    y_train, y_test = y[0:n_train], y[n_train:n_train+n_test]
+    gbdt1 = HistGradientBoostingRegressor(loss='poisson', random_state=0)
```
```suggestion
    gbdt1 = HistGradientBoostingRegressor(loss='poisson', random_state=rng)
```
For consistency. Though that won't matter: the RNG is only relevant for early stopping which won't be activated.

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_gradient_boosting.py:
```diff
@@ -192,6 +195,56 @@ def test_least_absolute_deviation():
+    gbdt2.fit(X_train, y_train)
+    dummy = DummyRegressor(strategy="mean").fit(X_train, y_train)
+
+    # Loss on training set
```
Maybe merge both parts with a simple for loop

`for X in (X_train, X_test): ...`

NicolasHug commented on sklearn/ensemble/_hist_gradient_boosting/loss.py:
```diff
@@ -269,6 +273,62 @@ def update_leaves_values(self, grower, y_true, raw_predictions,
+        # TODO: For speed, we could remove the constant xlogy(y_true, y_true)
+        # Advantage of this form: minimum of zero at raw_predictions = y_true.
+        loss = (xlogy(y_true, y_true) - y_true * (raw_predictions + 1)
+                + np.exp(raw_predictions))
```
Should there be a .5 somewhere?
(I assume not, otherwise the gradient checks would fail, just want to make sure)

lorentzenchr replied:
https://github.com/scikit-learn/scikit-learn/blob/2a5d40f5d102d19fd874134a2443ec0cdc7e0cdc/sklearn/_loss/glm_distribution.py#L314-L315
`1/2 * poisson_deviance = 1/2 * 2 * (xlog(y_true, y_true/y_pred) - y_true + y_pred = xlogy(y_true, y_true) - y_true*log(exp(raw)) - y_true + exp(raw)`
Pretty sure:+1:

NicolasHug reviewed (approved):
Thanks @lorentzenchr , a few more nits but looks good!

Thanks for the fast work!

(I wonder why the CI doesn't show the test suite instances... tests pass locally at least)

pinging @ogrisel who will be interested.

Commit:
DOC whats new wording

Commit:
MNT remove wrong comment

Commit:
MNT nitpicks

Commit:
Merge branch 'master' into hgb_poisson

lorentzenchr commented:
Can someone explain, why the test suddenly fails and how to resolve it?
```
___ test_estimators[HistGradientBoostingRegressor()-check_estimators_unfitted] ____
...
def predict(self, X):
...
>       return self.loss_.inverse_link_function(self._raw_predict(X).ravel())
E       AttributeError: 'HistGradientBoostingRegressor' object has no attribute 'loss_'
```
An unfitted `HistGradientBoostingRegressor`does not have attribute `loss_` only attribute `loss`.

NicolasHug commented:
You'll need to call `check_is_fitted` in `predict` of `HistGradientBoostingRegressor` now.

I think the error comes from the fact that you added 

`        return self.loss_.inverse_link_function(self._raw_predict(X).ravel())`

so now the error is "this estimator doesn't have a loss_ attribute" instead of being "this estimator isn't fitted" (as would be raised by `_raw_predict()`)

Alternatively this should also work

```py
		pred = self._raw_predict(X).ravel()
        return self.loss_.inverse_link_function(pred)
```

Commit:
BUG check if fitted in predict of hist gradient boosting

lorentzenchr commented:
@NicolasHug Thanks. That solves it. In particular, I was wondering why the tests did pass before. Never mind.

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/loss.py:
```diff
@@ -265,6 +269,59 @@ def update_leaves_values(self, grower, y_true, raw_predictions,
+        # shape (1, n_samples) --> (n_samples,). reshape(-1) is more likely to
+        # return a view.
+        raw_predictions = raw_predictions.reshape(-1)
+        target = np.empty_like(raw_predictions, dtype=Y_DTYPE)
+        target = np.exp(raw_predictions)
```
```suggestion
        target = np.exp(raw_predictions, dtype=Y_DTYPE)
```

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/loss.py:
```diff
@@ -265,6 +269,59 @@ def update_leaves_values(self, grower, y_true, raw_predictions,
+        # return a view.
+        raw_predictions = raw_predictions.reshape(-1)
+        # TODO: For speed, we could remove the constant xlogy(y_true, y_true)
+        # Advantage of this form: minimum of zero at raw_predictions = y_true.
```
Are we taking advantage of this advantage somewhere?

lorentzenchr replied:
Not that I know of. Might be interesting to see, if it matters (at all).

thomasjpfan reviewed (commented):
Nice work here @lorentzenchr

Commit:
MNT remove unnecessary np.empty_like

thomasjpfan commented on sklearn/ensemble/_hist_gradient_boosting/tests/test_loss.py:
```diff
@@ -164,6 +171,27 @@ def test_baseline_least_absolute_deviation():
+    # log(0)
+    assert y_train.sum() > 0
+    baseline_prediction = loss.get_baseline_prediction(y_train, None, 1)
+    assert baseline_prediction.shape == tuple()  # scalar
```
Nit:

```suggestion
    assert np.isscaler(baseline_prediction)
```

Commit:
address review comments for tests

Commit:
remove unused method Poisson.predict_target

thomasjpfan commented:
Thank you @lorentzenchr !

lorentzenchr commented:
@thomasjpfan Thank you for your review and merging. :+1:

Now, the good old, brand new Poisson GLM will come out in the same release as this Poisson HGB. That is a strong competitor! 😄

rth commented:
Thanks for all the work by the three of you in this PR! Looking forward to the release :)
